pkg/controller/pushapplication/pushapplication_controller.go

```diff
@@ -81,6 +82,17 @@ func (r *ReconcilePushApplication) Reconcile(request reconcile.Request) (reconci
 		return reconcile.Result{}, err
 	}
 
+	// Check if the namespace of the cr is in the APP_NAMESPACES environment variable provided to the operator or if it's in the operator namespace
+	operatorNamespace, err := k8sutil.GetOperatorNamespace()
```

Comment:
Does that `k8sutil.GetOperatorNamespace()` return something valid when it's run locally? If not, does it return an error?

Reply:
Didn't really try it but I don't think it would work when running locally:

```
// GetOperatorNamespace returns the namespace the operator should be running in.
func GetOperatorNamespace() (string, error) {
	nsBytes, err := ioutil.ReadFile("/var/run/secrets/kubernetes.io/serviceaccount/namespace")
	if err != nil {
		if os.IsNotExist(err) {
			return "", ErrNoNamespace
		}
		return "", err
	}
	ns := strings.TrimSpace(string(nsBytes))
	log.V(1).Info("Found namespace", "Namespace", ns)
	return ns, nil
}
```

Reply:
I think we can use `k8sutil.GetWatchNamespace()` here as it is already used in `main.go`.

That can be used for sure.

https://github.com/aerogear/unifiedpush-operator/blob/f144d9ed14b0db7987a51d62dc2a31f2dc745547/cmd/manager/main.go#L67

Reply:
If we run the operator locally, that `k8sutil.GetOperatorNamespace()` will return with that `ErrNoNamespace` error, because the file it checks doesn't exist. This means that if we try to create a PushApplication in the unifiedpush namespace, it won't work.

We currently set the watch namespace to be an empty string so that the operator will watch more than one namespace (it needs to watch its own namespace, as that's where we want the UPS instance to be, and it also needs to watch the namespace(s) where PushApplication and Variant CRs will be created). When running `operator-sdk up local` we could add `--namespace unifiedpush`, and that will set the watch namespace to that value, but again, that will cause the operator to only listen in that one namespace I think. Maybe that's sufficient when running the operator locally?

Alternatively, if `k8sutil.GetOperatorNamespace()` returns an error, we could have it check some undocumented env var such as `_LOCAL_OPERATOR_NAMESPACE` or something, and pass that in the Makefile?

Reply:
I am confused. There are too many namespace contexts here :) Watch, operator, app, anything else? :)

Let me think out loud.

Let's go 1 step back.

* We want the UPS operator to run in unifiedpush namespace
* We want the UPS operator to create UPS server in unifiedpush namespace
* We want the UPS operator to expect the CRs in `APP_NAMESPACES`
** Does it also include the installation CR? Or is it only for service consumption CRs?
* We want he UPS operator to watch for pods/deployments/etc. in unifiedpush namespace.

We already have a separate env var for `APP_NAMESPACES`.
For all the others, we just need to use unifiedpush.
* Running in cluster: `k8sutil.GetOperatorNamespace()` should work.
* Running locally: Using `WATCH_NAMESPACE` can be enforced in the makefile.

So, here:

https://github.com/aerogear/unifiedpush-operator/pull/42#discussion-diff-302098737R86

 we can do like:
* Get `k8sutil.GetOperatorNamespace()`.
* If it fails, get `k8sutil.GetWatchNamespace()`
* If it is empty, `panic`

wdyt?

Reply:
My understanding of the problem is that:

* We have `APP_NAMESPACES` because we want to limit the namespaces that we react to consumption CRs in, but we also want to react to them in the operator namespace, whether it's part of `APP_NAMESPACES` or not (a user should be able to do everything with this operator in one namespace if they want). To answer your other question about `APP_NAMESPACES`, it's only in the "consumption" controllers that we want to check this in (the whole idea of it is a dirty hack, since the operator will react to "consumption" CRs in any namespace, but will just refuse to do anything if it's not in an acceptable namespace).

* We need to set `WATCH_NAMESPACE` to an empty string to generally allow listening in other namespaces than the operator one.

So the problem with your proposed steps is that if the `WATCH_NAMESPACE` env var is not an empty string, and you have anything other than the same single namespace in `APP_NAMESPACES`, those additional namespaces will be ignored.

Maybe it's fine for now that for _local_ development, that we disallow creation of "consumption" CRs (not sure how I feel about that term), in the operator namespace, until there's a better solution from the operator-sdk (I think there will be)? What do you think? :thinking:

Reply:
> "consumption" CRs (not sure how I feel about that term)

LOL :) I didn't make it up, somebody else did :grin: 

> What do you think? 

Sounds good. Let's proceed with this.

Reply:
>> "consumption" CRs (not sure how I feel about that term)
>
> LOL :) I didn't make it up, somebody else did :grin: 

I know, and I use it sometimes too; it was just in my third usage of it in that comment I realised I'd had enough of it :sweat_smile: 

>> What do you think?
>
> Sounds good. Let's proceed with this.

Cool, I think that means that this PR is already as expected. I'll just double-check.